src/test/scala/sigmastate/lang/SigmaParserTest.scala

```diff
@@ -531,18 +532,179 @@ class SigmaParserTest extends PropSpec with PropertyChecks with Matchers with La
+          ApplyTypes(Ident("Coll", NoType), Vector(SBoolean)),
+          Vector(TrueLeaf, FalseLeaf))
+      ))
+  }
+
+  property("SBoolean.toByte") {
+    parse("true.toByte") shouldBe Select(TrueLeaf, "toByte")
+  }
+
+  property("SOption.map") {
+    parse("Some(1).map { (b: Int) => b}") shouldBe
+      Apply(Select(Apply(Ident("Some", NoType), Vector(IntConstant(1))), "map", None),
+        Vector(Lambda(List(), Vector(("b", SInt)), NoType, Some(Ident("b", NoType)))))
+  }
+
+  property("SCollection.zip") {
+    parse("OUTPUTS.zip(Coll(1, 2))") shouldBe
+      Apply(Select(Ident("OUTPUTS"), "zip"),
+        Vector(Apply(Ident("Coll"), Vector(IntConstant(1), IntConstant(2)))))
+  }
+
+  property("SCollection.zipWith") {
+    parse("OUTPUTS.zipWith(Coll(1, 2), { (box: Box, i: Int) => i })") shouldBe
+      Apply(Select(Ident("OUTPUTS"), "zipWith"),
+        Vector(Apply(Ident("Coll"), Vector(IntConstant(1), IntConstant(2))),
+          Lambda(List(), Vector(("box", SBox), ("i", SInt)), NoType, Some(Ident("i", NoType)))))
+  }
+
+  property("SCollection.flatMap") {
+    parse("OUTPUTS.flatMap({ (box: Box) => Coll(box) })") shouldBe
+      Apply(Select(Ident("OUTPUTS"), "flatMap"),
+        Vector(Lambda(List(), Vector(("box", SBox)), NoType,
+          Some(Apply(Ident("Coll"), Vector(Ident("box", NoType)))))))
+  }
+
+  property("SGroupElement.exp") {
+    parse("{ (g: GroupElement) => g.exp(1.toBigInt) }") shouldBe
+      Lambda(List(), Vector(("g", SGroupElement)), NoType,
+        Some(Apply(Select(Ident("g", NoType), "exp", None),
+          Vector(Select(IntConstant(1), "toBigInt", None)))))
+  }
+
+  property("SNumeric.toBytes") {
+    parse("1.toBytes") shouldBe Select(IntConstant(1), "toBytes")
+    parse("1L.toBytes") shouldBe Select(LongConstant(1), "toBytes")
+  }
+
+  property("SNumeric.toBits") {
+    parse("1.toBits") shouldBe Select(IntConstant(1), "toBits")
+    parse("1L.toBits") shouldBe Select(LongConstant(1), "toBits")
+  }
+
+  property("SNumeric.abs") {
+    parse("1.abs") shouldBe Select(IntConstant(1), "abs")
+  }
+
+  property("SNumeric.compare") {
+    parse("1.compare(2)") shouldBe Apply(Select(IntConstant(1), "compare", None), Vector(IntConstant(2)))
+  }
+
+  property("numeric constant negation") {
+    parse("-3") shouldBe IntConstant(-3)
+    parse("-3.toByte") shouldBe Select(IntConstant(-3), "toByte")
+  }
+
+  property("numeric negation unary op") {
+    parse("-OUTPUTS.size") shouldBe Negation(Select(Ident("OUTPUTS"), "size").asIntValue)
+    parse("- (3 - 2)") shouldBe Negation(Minus(IntConstant(3), IntConstant(2)))
+  }
+
+  property("bitwise inversion unary op ~") {
+    parse("~OUTPUTS.size") shouldBe BitInversion(Select(Ident("OUTPUTS"), "size").asIntValue)
+  }
+
+  property("HEADERS methods") {
+    parse("HEADERS") shouldBe Ident("HEADERS")
+    parse("HEADERS.version") shouldBe Select(Ident("HEADERS"), "version")
+    parse("HEADERS.parentId") shouldBe Select(Ident("HEADERS"), "parentId")
```

Comment:
@greenhat Well, parser doesn't check it, but `HEADERS` is of `Coll` type

Reply:
Thanks! Missed that one. Fixed.